src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -102,6 +110,43 @@ public boolean reduceScopeForLimitedSpace(Set<SSTableReader> nonExpiredSSTables,
         return false;
     }
 
+    private void maybeNotifyIndexersAboutRowsInFullyExpiredSSTables(Set<SSTableReader> fullyExpiredSSTables)
+    {
+        if (fullyExpiredSSTables.isEmpty())
+            return;
+
+        List<Index> indexes = cfs.indexManager.listIndexes()
+                                              .stream()
+                                              .filter(Index::notifyIndexerAboutRowsInFullyExpiredSSTables)
+                                              .collect(Collectors.toList());
+
+        if (indexes.isEmpty())
+            return;
+
+        for (SSTableReader expiredSSTable : fullyExpiredSSTables)
+        {
+            expiredSSTable.getScanner()
+                          .forEachRemaining(partition ->
+                                            {
+                                                partition.forEachRemaining(unfiltered -> {
+                                                    if (unfiltered instanceof Row)
+                                                    {
+                                                        CompactionTransaction transaction = cfs.indexManager.newCompactionTransaction(partition.partitionKey(),
+                                                                                                                                      partition.columns(),
+                                                                                                                                      1,
+                                                                                                                                      FBUtilities.nowInSeconds(),
+                                                                                                                                      // only indexes which matter
+                                                                                                                                      indexes);
```

Comment:
indexes which matter only, which have `notifyIndexerAboutRowsInFullyExpiredSSTables` to be true. I have accommodated `newCompactionTransaction` to pass indexes explicitly.